include/os/freebsd/zfs/sys/zfs_znode_impl.h

```diff
@@ -121,29 +121,48 @@ typedef struct zfs_soft_state {
 #define	zn_rlimit_fsize(zp, uio) \
     vn_rlimit_fsize(ZTOV(zp), GET_UIO_STRUCT(uio), zfs_uio_td(uio))
 
-#define	ZFS_ENTER_ERROR(zfsvfs, error) do {			\
-	ZFS_TEARDOWN_ENTER_READ((zfsvfs), FTAG);		\
-	if (__predict_false((zfsvfs)->z_unmounted)) {		\
-		ZFS_TEARDOWN_EXIT_READ(zfsvfs, FTAG);		\
-		return (error);					\
-	}							\
-} while (0)
-
 /* Called on entry to each ZFS vnode and vfs operation  */
-#define	ZFS_ENTER(zfsvfs)	ZFS_ENTER_ERROR(zfsvfs, EIO)
+static inline int
+zfs_enter(zfsvfs_t *zfsvfs, char *tag)
+{
+	ZFS_TEARDOWN_ENTER_READ(zfsvfs, tag);
+	if (__predict_false((zfsvfs)->z_unmounted)) {
+		ZFS_TEARDOWN_EXIT_READ(zfsvfs, tag);
+		return (EIO);
+	}
+	return (0);
+}
 
 /* Must be called before exiting the vop */
-#define	ZFS_EXIT(zfsvfs)	ZFS_TEARDOWN_EXIT_READ(zfsvfs, FTAG)
-
-#define	ZFS_VERIFY_ZP_ERROR(zp, error) do {			\
-	if (__predict_false((zp)->z_sa_hdl == NULL)) {		\
-		ZFS_EXIT((zp)->z_zfsvfs);			\
-		return (error);					\
-	}							\
-} while (0)
+static inline void
+zfs_exit(zfsvfs_t *zfsvfs, char *tag)
+{
+	ZFS_TEARDOWN_EXIT_READ(zfsvfs, tag);
+}
 
 /* Verifies the znode is valid */
-#define	ZFS_VERIFY_ZP(zp)	ZFS_VERIFY_ZP_ERROR(zp, EIO)
+#define	zfs_verify_zp(zp)				\
+/* CSTYLED */						\
+({							\
+	int ___err = 0;					\
+	if (__predict_false((zp)->z_sa_hdl == NULL))	\
+		___err = EIO;				\
+	___err;						\
+})
+
+/* zfs_enter and zfs_verify_zp together */
+#define	zfs_enter_verify_zp(zfsvfs, zp, tag)	\
```

Comment:
@behlendorf It's not possible because znode_t is not yet defined.
The only way to make it happen is to move it to a different file.
We can move it to the common zfs_znode.h, but there's the difference of __predict_false and unlikely.
Do we have a abstraction these two?

Reply:
`unlikely()` and `__predict_false()` are defined pretty much the same way, you can use `unlikely()`.